### ISSUE TITLE
Can RaySGD with Pytorch share Object Store when multiple GPUs are executed?

### ISSUE THREAD

**Issue by Matrix-World** [ray-project/ray#13929]
Hello,
I followed the MNIST Tutorial to train an MNIST classifier with Pytorch Lightning's LightningModule and Ray SGD (https://docs.ray.io/en/master/raysgd/raysgd_ptl.html). But currently we still mainly use Torch, so we had trained an MNIST classifier with Pytorch(not pytorch_lightning) and Ray SGD, Just convert LightningModule to nn.Module and rewrite train_epoch and validate methods( case A). In another case, only pytorch(DDP) is used to implement MNIST classifier without using RaySGD(case B). Both cases are DistributedDataParallel.
In the train_epoch method, the following code block will be executed:
--------edite line---------
output = mode(*features)
loss = F.nll_loss(output,target)
print(loss.device)
optimizer.zero_grad()
loss.backward()
optimizer.step()
-------edite line----------
 Cases A and B both use 4 GPUs (no cross-nodes), but it is very strange that the GPU label of the loss output by RaySGD with Pytorch(case A) are all cuda:0 (check with nvidia-smi to find GPU: 2, 3, 4, 5 in using). When using only Pytorch(case B), the output loss GPU label are cuda:2, cuda:3, cuda:4, cuda:5 (check with nvidia-smi to find that GPUs: 2, 3, 4, 5 are in using).
It seems that cuda:0 is used as a shared memory (object-store) in case A. During the weight update process, it seems that other GPU processes are getting loss at cuda:0. So can you explain the All-Reduce process?
Looking forward to your reply~

**Comment by amogkam** [comment]
Ray automatically sets the "CUDA_VISIBLE_DEVICES" environment variable for all remote workers (more info [here](https://docs.ray.io/en/master/using-ray-with-gpus.html)). So with RaySGD, each worker has 1 gpu and the device is `cuda:0`. But the 0 actually indexes into "CUDA_VISIBLE_DEVICES", which is different for each worker. You can confirm this by printing out this environment variable in `train_epoch`.

**Comment by Matrix-World** [comment]
> Ray automatically sets the "CUDA_VISIBLE_DEVICES" environment variable for all remote workers (more info [here](https://docs.ray.io/en/master/using-ray-with-gpus.html)). So with RaySGD, each worker has 1 gpu and the device is `cuda:0`. But the 0 actually indexes into "CUDA_VISIBLE_DEVICES", which is different for each worker. You can confirm this by printing out this environment variable in `train_epoch`.

Thank you, I get it
